platform-common-string/src/test/java/com/softicar/platform/common/string/csv/CsvTokenizerTest.java

```diff
@@ -0,0 +1,285 @@
+
+		tokenize("qwe,asd,zxc\n123,456,789");
+		new Asserter()//
+			.assertRow("qwe", "asd", "zxc")
+			.assertRow("123", "456", "789")
+			.assertNoMoreRows();
+	}
+
+	@Test
+	public void testWithMultipleLinesSeparatedByBackslashR() {
+
+		tokenize("qwe,asd,zxc\r123,456,789");
+		new Asserter()//
+			.assertRow("qwe", "asd", "zxc")
+			.assertRow("123", "456", "789")
+			.assertNoMoreRows();
+	}
+
+	@Test
+	public void testWithMultipleLinesSeparatedByBackslashRN() {
+
+		tokenize("qwe,asd,zxc\r\n123,456,789");
+		new Asserter()//
+			.assertRow("qwe", "asd", "zxc")
+			.assertRow("123", "456", "789")
+			.assertNoMoreRows();
+	}
+
+	@Test
+	public void testWithMultipleLinesAndEmptyLine() {
+
+		tokenize("qwe,asd,zxc\n\n123,456,789");
+		new Asserter()//
+			.assertRow("qwe", "asd", "zxc")
+			.assertRow("123", "456", "789")
+			.assertNoMoreRows();
+	}
+
+	@Test
+	public void testWithMultipleLinesOfVariableLength() {
+
+		tokenize("qwe,asd,zxc\n123,456\nxxx,yyy,zzz,___");
+		new Asserter()//
+			.assertRow("qwe", "asd", "zxc")
+			.assertRow("123", "456")
+			.assertRow("xxx", "yyy", "zzz", "___")
+			.assertNoMoreRows();
+	}
+
+	// ---------------- value separation ---------------- //
+
+	@Test
+	public void testWithValueSeparationByAdjacentCommas() {
+
+		tokenize(",,qwe,asd,, ,zxc,,");
+		new Asserter()//
+			.assertRow("", "", "qwe", "asd", "", " ", "zxc", "", "")
+			.assertNoMoreRows();
+	}
+
+	@Test
+	public void testWithValueSeparationMissing() {
+
+		try {
+			tokenize("\"qwe\" \"asd\",zxc");
+		} catch (CsvFormatException exception) {
+			assertEquals(
+				"Invalid CSV format after character 5: After a double quoted value, a comma, a newline or end-of-file are expected.",
+				exception.getMessage());
+			return;
+		}
+		fail();
+	}
+
+	// ---------------- line-spanning value ---------------- //
+
+	@Test
+	public void testWithLineSpanningValue() {
+
+		tokenize("\"X\nX\r\nX\rX\",asd,zxc");
+		new Asserter()//
+			.assertRow("X\nX\r\nX\rX", "asd", "zxc")
+			.assertNoMoreRows();
+	}
+
+	// ---------------- spacing ---------------- //
+
+	@Test
+	public void testWithSpacesAroundValues() {
+
+		tokenize(" qwe, asd ,zxc ");
+		new Asserter()//
+			.assertRow(" qwe", " asd ", "zxc ")
+			.assertNoMoreRows();
+	}
+
+	// ---------------- quoting ---------------- //
+
+	@Test
+	public void testWithQuotes() {
+
+		tokenize("\"\"\"qwe,123\"\"\",\"asd\",zxc");
+		new Asserter()//
+			.assertRow("\"qwe,123\"", "asd", "zxc")
+			.assertNoMoreRows();
+	}
+
+	@Test
+	public void testWithQuotesOnEdges() {
+
+		tokenize("\"qwe\",asd,\"zxc\"");
+		new Asserter()//
+			.assertRow("qwe", "asd", "zxc")
+			.assertNoMoreRows();
+	}
+
+	@Test
+	public void testWithQuotesMissingInTheEnd() {
+
+		try {
+			tokenize("qwe,\"asd,zxc");
+		} catch (CsvFormatException exception) {
+			assertEquals("Invalid CSV format after character 3: Expected a double quote at the end of the value.", exception.getMessage());
```

Comment:
Yes, there is in error _somewhere after_ character 3 (after "qwe"). But it would be better if the error message is more exact.
In this case there must be another double quote after character 12.

Reply:
`CsvTokenized` was rewritten from scratch and should now yield precise syntax error coordinates. Pls check it out.